### ISSUE TITLE
`method_missing': undefined method `path' for [class] (NoMethodError)

### ISSUE THREAD

**Issue by iandonovan** [RickCarlino/smarf_doc#3]
Hey Rick!

[Here's the line that barfs](https://github.com/RickCarlino/smarf_doc/blob/master/lib/base.rb#L10), and I'm fairly certain allowing nil values in the sort [like this](http://stackoverflow.com/a/808694) will do the trick, I don't think that'd fix the root of the problem.

What are your thoughts? I can get a PR up that implements the above sort if you'd like.

**Comment by RickCarlino** [comment]
Hey pretzel buddy- let me take a look right now.

Hope you're doing well. :interlocked-pretzels:

**Comment by RickCarlino** [comment]
@iandonovan let me make sure I understand this correctly:

You are getting `nil` values because you're trying to make a request to the root URL? How is it possible that `request.path` could be `nil`? I've never come across that issue while using it before. Do you think you could show me an example spec that will trigger this?

Obviously, we could fix it by doing something like this:

```
  def sort_by_url!
    @tests.sort! do |x, y|
      (x.request.path || '') <=> (y.request.path || '')
    end
  end
```

we could also call `compact` on [this line](https://github.com/RickCarlino/smarf_doc/blob/master/lib/base.rb#L59) but I'm really curious what caused it in the first place. Nil checking give me asthma attacks.

**Comment by iandonovan** [comment]
![](http://www.auntieannes.com/public/images/pretzels-nutrition/soft-pretzels@2x.png)

Here's a more-complete stack trace that I've scrubbed (I'm sure you understand):

``` ruby
/Users/USER_NAME/.rvm/gems/ruby-2.1.3/gems/activemodel-4.2.0/lib/active_model/attribute_methods.rb:433:in `method_missing': undefined method `path' for #<CLASS_NAME:0x007feb3a25aa08> (NoMethodError)
    from /Users/USER_NAME/.rvm/gems/ruby-2.1.3/bundler/gems/smarf_doc-7c01c88e6d18/lib/base.rb:10:in `block in sort_by_url!'
    from /Users/USER_NAME/.rvm/gems/ruby-2.1.3/bundler/gems/smarf_doc-7c01c88e6d18/lib/base.rb:9:in `sort!'
    from /Users/USER_NAME/.rvm/gems/ruby-2.1.3/bundler/gems/smarf_doc-7c01c88e6d18/lib/base.rb:9:in `sort_by_url!'
    from /Users/USER_NAME/.rvm/gems/ruby-2.1.3/bundler/gems/smarf_doc-7c01c88e6d18/lib/base.rb:59:in `finish!'
    from /Users/USER_NAME/Work/PROJECT_NAME/spec/spec_helper.rb:67:in `block (2 levels) in <top (required)>'
```

Here's the thing! It fails as part of the `config.after(:suite) { SmarfDoc.finish! }` hook in `spec_helper`. It yells about one particular instance of a model -- not a controller, but the model -- not having a `path` method. So, somehow, `x.request` is being set as an instance of the model.

If I run the corresponding model's unit and functional specs individually, I get no failure. I only ever see this exception when I run the full test suite with `rake`.

How is `request` set?

**Comment by RickCarlino** [comment]
Can you show me what your `SmarfDoc.run!` call(s) look like?

**Comment by iandonovan** [comment]
I fixed it!

I threw a debugger before that `SmarfDoc.run!` call and saw that, indeed, `request` was being set as an instance of the model, not the typical `request` object. Why was that? Because of the `let(:request) { create(:request) }` line directly above the spec. The model's name is has "-Request" in it, so I went ahead and named the variable `request`, which is a reserved word.

I'm surprised the tests passed, actually, but that's an investigation for another day. Anyway, changing the variable name to `req` fixed my problem.

Great gem, btw.
